meshtastic/slog/slog.py

```diff
@@ -0,0 +1,208 @@
+class LogDef:
+    """Log definition."""
+
+    code: str  # i.e. PM or B or whatever... see meshtastic slog documentation
+    format: parse.Parser  # A format string that can be used to parse the arguments
+
+    def __init__(self, code: str, fmt: str) -> None:
+        """Initialize the LogDef object.
+
+        code (str): The code.
+        format (str): The format.
+        """
+        self.code = code
+        self.format = parse.compile(fmt)
+
+
+"""A dictionary mapping from logdef code to logdef"""
+log_defs = {
+    d.code: d
+    for d in [
+        LogDef("B", "{boardid:d},{version}"),
+        LogDef("PM", "{bitmask:d},{reason}"),
+    ]
+}
+log_regex = re.compile(".*S:([0-9A-Za-z]+):(.*)")
+
+
+class PowerLogger:
+    """Logs current watts reading periodically using PowerMeter and ArrowWriter."""
+
+    def __init__(self, pMeter: PowerMeter, file_path: str, interval=0.2) -> None:
+        """Initialize the PowerLogger object."""
+        self.pMeter = pMeter
+        self.writer = FeatherWriter(file_path)
+        self.interval = interval
+        self.is_logging = True
+        self.thread = threading.Thread(
+            target=self._logging_thread, name="PowerLogger", daemon=True
+        )
+        self.thread.start()
+
+    def _logging_thread(self) -> None:
+        """Background thread for logging the current watts reading."""
+        while self.is_logging:
+            d = {
+                "time": datetime.now(),
+                "average_mW": self.pMeter.get_average_current_mA(),
+                "max_mW": self.pMeter.get_max_current_mA(),
+                "min_mW": self.pMeter.get_min_current_mA(),
+            }
+            self.pMeter.reset_measurements()
+            self.writer.add_row(d)
+            time.sleep(self.interval)
+
+    def close(self) -> None:
+        """Close the PowerLogger and stop logging."""
+        if self.is_logging:
+            self.pMeter.close()
+            self.is_logging = False
+            self.thread.join()
+            self.writer.close()
+
+
+# FIXME move these defs somewhere else
+TOPIC_MESHTASTIC_LOG_LINE = "meshtastic.log.line"
+
+
+class StructuredLogger:
+    """Sniffs device logs for structured log messages, extracts those into apache arrow format.
+    Also writes the raw log messages to raw.txt"""
+
+    def __init__(self, client: MeshInterface, dir_path: str) -> None:
+        """Initialize the StructuredLogger object.
+
+        client (MeshInterface): The MeshInterface object to monitor.
+        """
+        self.client = client
+        self.writer = FeatherWriter(f"{dir_path}/slog")
+        self.raw_file: Optional[
+            io.TextIOWrapper
+        ] = open(  # pylint: disable=consider-using-with
+            f"{dir_path}/raw.txt", "w", encoding="utf8"
+        )
+
+        # We need a closure here because the subscription API is very strict about exact arg matching
+        def listen_glue(line, interface):  # pylint: disable=unused-argument
```

Comment:
(mostly noting to myself -- another spot that'll need updating in order to pass a LogRecord instead of just the message via pubsub)